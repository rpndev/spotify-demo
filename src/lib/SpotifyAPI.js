import axios from 'axios';
import {encode} from 'base-64';
import {
  SPOTIFY_API_PREFIX,
  SPOTIFY_CLIENT_ID,
  SPOTIFY_CLIENT_SECRET,
} from 'react-native-dotenv';

const SpotifyAPI = axios.create({
  baseURL: SPOTIFY_API_PREFIX,
});

SpotifyAPI.interceptors.response.use(
  res => res,
  async e => {
    console.log(e);
    if (e.status.status === 401) {
      await getAccessToken();
    }
  },
);

export const getAccessToken = async () => {
  const data = new URLSearchParams();
  data.append('grant_type', 'client_credentials');
  try {
    const res = await axios.post(
      'https://accounts.spotify.com/api/token',
      data,
      {
        headers: {
          Authorization: `Basic ${encode(
            SPOTIFY_CLIENT_ID + ':' + SPOTIFY_CLIENT_SECRET,
          )}`,
          'Content-Type': 'application/x-www-form-urlencoded',
        },
      },
    );
    SpotifyAPI.defaults.headers.Authorization = `Bearer ${res.data.access_token}`;
  } catch (e) {
    console.log('get access token failed: ', e);
  }
};

export default SpotifyAPI;
