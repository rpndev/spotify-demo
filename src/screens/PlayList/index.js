/**
 * @format
 * @flow
 */

import React, {useState, useEffect} from 'react';
import {
  View,
  Text,
  FlatList,
  Image,
  StyleSheet,
  TouchableOpacity,
} from 'react-native';
import {getCountry} from 'react-native-localize';
import SpotifyAPI from '../../lib/SpotifyAPI';

const PlayList = props => {
  const [playlist, setPlaylist] = useState([]);
  useEffect(() => {
    SpotifyAPI.get('/browse/featured-playlists', {
      params: {country: getCountry()},
    })
      .then(res => setPlaylist(res.data.playlists.items))
      .catch(e => console.log(e.response));
  }, []);

  return (
    <FlatList
      data={playlist}
      keyExtractor={item => item.id}
      renderItem={({item}) => (
        <TouchableOpacity
          onPress={() =>
            props.navigation.navigate('tracks', {playlist_id: item.id})
          }>
          <View style={styles.listItem}>
            <Image
              source={{uri: item.images[0].url, height: 60, width: 100}}
              resizeMode="cover"
              style={styles.listItemImage}
            />
            <View style={styles.listItemContent}>
              <Text style={styles.nameText}>{item.name}</Text>
              <Text>Number of tracks: {item.tracks.total}</Text>
            </View>
          </View>
        </TouchableOpacity>
      )}
      ItemSeparatorComponent={() => <View style={styles.seperator} />}
      style={styles.container}
      contentContainerStyle={styles.content}
    />
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'pink',
  },
  content: {
    padding: 30,
  },
  seperator: {
    height: 20,
  },
  listItem: {
    flexDirection: 'row',
    borderRadius: 10,
    backgroundColor: 'white',
    overflow: 'hidden',
    alignItems: 'center',
  },
  listItemContent: {
    flex: 1,
  },
  listItemImage: {
    borderRadius: 10,
    marginRight: 20,
  },
  nameText: {
    flexWrap: 'wrap',
    fontWeight: 'bold',
    fontSize: 14,
  },
});

export default PlayList;
