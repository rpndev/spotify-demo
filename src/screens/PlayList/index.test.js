import React from 'react';
import PlayList from './index';

import renderer from 'react-test-renderer';

import SpotifyAPI from '../../lib/SpotifyAPI';

function PlayListData({ id, image, name, traks }) {
  return {
    id,
    images: { url: image },
    name,
    traks: { total: traks }
  };
}

const stubbedAPIResponse = {
  playlists: {
    items: [
      PlayListData({
        id: 'dsflkhasfd3',
        image: 'https://i.scdn.co/image/ab67706f00000002487cfbde23e0179f7bcf37d5',
        name: 'Fresh Finds Presents... Class of 2019',
        traks: 100
      }),
      PlayListData({
        id: 'dsfgfd67sdfg76',
        image: 'https://i.scdn.co/image/ab67706f00000002487cfbde23e0179f7bcf37d5',
        name: 'Fresh Finds: Fire Emoji',
        traks: 120
      })
    ]
  }
};


beforeEach(() => {
    SpotifyAPI.get = jest.fn(() => Promise.resolve({ data: { data: stubbedAPIResponse }}));
});

describe('PlayList', () => {
  it('renders correctly after fetching API', () => {
    const props = { 
      navigation: {
        navigate: jest.fn()
      }
    };

    const tree = renderer.create(<PlayList {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});