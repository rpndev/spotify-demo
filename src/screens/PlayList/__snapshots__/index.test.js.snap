// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingScreen renders correctly Loading Screen 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  contentContainerStyle={
    Object {
      "padding": 30,
    }
  }
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "backgroundColor": "pink",
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View />
</RCTScrollView>
`;

exports[`PlayList renders correctly after fetching API 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  contentContainerStyle={
    Object {
      "padding": 30,
    }
  }
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "backgroundColor": "pink",
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View />
</RCTScrollView>
`;
